File: tests/imo3d-depth-architecture.test.py
import importlib.util
from pathlib import Path
import unittest
spec=importlib.util.spec_from_file_location("architecture",Path(__file__).resolve().parents[1]/"scripts/imo3d-depth-architecture.py")
core=importlib.util.module_from_spec(spec);spec.loader.exec_module(core);np=core.np

def scenes():return [{"id":"a","position":{"x":-.3,"y":0,"z":0},"floor":0},{"id":"b","position":{"x":.3,"y":0,"z":0},"floor":0}]
def wall(a,b,index="w"):
    return {"id":index,"normal":{"x":0.,"z":1.},"offset":-2.,"segments":[{"a":{"x":a[0],"z":a[1]},"b":{"x":b[0],"z":b[1]}}]}

class ArchitectureEvidence(unittest.TestCase):
    def test_repeated_crop_ids_do_not_count_as_multiple_cameras(self):
        samples=[{"x":0,"y":0,"z":-2,"confidence":.9,"sceneIds":["a","a","a"]}]
        points,_,_,rejected,_=core.prepare_samples(samples,scenes());self.assertEqual(len(points),0);self.assertEqual(rejected,1)

    def test_raw_network_confidence_cannot_be_mistaken_for_normalized_support(self):
        samples=[{"x":0,"y":0,"z":-2,"confidence":8.25,"sceneIds":["a","b"]}]
        self.assertEqual(len(core.prepare_samples(samples,scenes())[0]),0)

    def test_missing_wall_does_not_become_convex_closure(self):
        walls=[wall((-2,-2),(2,-2),"south"),wall((2,-2),(2,2),"east"),wall((2,2),(-2,2),"north")]
        self.assertEqual(core.observed_room_cycles(walls),[])
        walls.append(wall((-2,2),(-2,-2),"west"));result=core.observed_room_cycles(walls)
        self.assertEqual(len(result),1);self.assertAlmostEqual(result[0]["areaCameraHeightSquared"],16)

    def test_unsupported_corner_extensions_and_uninhabited_boxes_are_not_rooms(self):
        walls=[wall((-2,-2),(1.5,-2),"south"),wall((2,-2),(2,2),"east"),wall((2,2),(-2,2),"north"),wall((-2,2),(-2,-2),"west")]
        self.assertEqual(core.observed_room_cycles(walls),[])
        walls[0]=wall((-2,-2),(2,-2),"south")
        self.assertEqual(core.observed_room_cycles(walls,camera_centers=[np.array([4,4])]),[])

    def test_no_points_in_wall_gap_is_not_a_doorway(self):
        candidate=wall((-2,-2),(-.4,-2));candidate["segments"]+=wall((.4,-2),(2,-2))["segments"]
        points=np.array([[x,y,-1] for x in np.linspace(-.5,.5,12) for y in np.linspace(-.8,.2,12)])
        self.assertEqual(core.infer_openings([candidate],points,[{"a","b"} for _ in points],scenes(),-1),[])

    def test_door_void_requires_rays_from_two_cameras_and_height_coverage(self):
        candidate=wall((-2,-2),(-.4,-2));candidate["segments"]+=wall((.4,-2),(2,-2))["segments"]
        points=np.array([[x,y,-3] for x in np.linspace(-.4,.4,12) for y in np.linspace(-.8,.2,12)])
        output=core.infer_openings([candidate],points,[{"a","b"} for _ in points],scenes(),-1)
        self.assertEqual(len(output),1);self.assertFalse(output[0]["verified"]);self.assertEqual(output[0]["widthInterpretation"],"visible_span_lower_bound")
        self.assertEqual(core.infer_openings([candidate],points,[{"a"} for _ in points],scenes(),-1),[])

    def test_upper_window_rays_and_closed_door_returns_do_not_create_walkthrough_apertures(self):
        candidate=wall((-2,-2),(-.4,-2));candidate['segments']+=wall((.4,-2),(2,-2))['segments']
        upper=np.array([[x,y,-3] for x in np.linspace(-.4,.4,12) for y in np.linspace(.4,1.2,12)])
        closed=np.array([[x,y,-2] for x in np.linspace(-.4,.4,12) for y in np.linspace(-.8,.8,12)])
        for points in [upper,closed]:self.assertEqual(core.infer_openings([candidate],points,[{'a','b'} for _ in points],scenes(),-1),[])

    def test_floor_and_low_furniture_do_not_become_vertical_room_walls(self):
        points=np.array([[x,y,-2] for x in np.arange(-2,2,.045) for y in np.arange(-.95,-.35,.045)])
        normal=np.tile([0.,0.,1.],(len(points),1));planar=np.ones(len(points),bool)
        result=core.fit_wall_planes(points,[{"a","b"} for _ in points],np.ones(len(points))*.9,normal,planar,-1)
        self.assertEqual(result,[])

    def test_observed_tall_multicamera_surface_has_supported_extent(self):
        points=np.array([[x,y,-2] for x in np.arange(-2,2,.045) for y in np.arange(-.95,.8,.045)])
        normal=np.tile([0.,0.,1.],(len(points),1));planar=np.ones(len(points),bool)
        result=core.fit_wall_planes(points,[{"a","b"} for _ in points],np.ones(len(points))*.9,normal,planar,-1)
        self.assertEqual(len(result),1);self.assertAlmostEqual(abs(result[0]["offset"]),2);self.assertEqual(result[0]["supportCameras"],["a","b"])

    def test_independent_floors_or_components_are_not_merged(self):
        mixed=scenes();mixed[1]["floor"]=1
        with self.assertRaises(ValueError):core.reconstruct_architecture({"scenes":mixed,"pointSamples":[]})
        mixed=scenes();mixed[0]["componentId"]="one";mixed[1]["componentId"]="two"
        with self.assertRaises(ValueError):core.reconstruct_architecture({"scenes":mixed,"pointSamples":[]})

    def test_bounded_refinement_keeps_identity_apertures_and_unsupported_lines(self):
        room={'id':'room','outline':[{'x':-2,'z':-2},{'x':2,'z':-2},{'x':2,'z':2},{'x':-2,'z':2}],'openings':[0],'doorwayCandidates':[{'edge':1,'offset':.4,'width':.1}]}
        west=wall((-2.1,-2),(-2.1,2),'west');west.update(normal={'x':1.,'z':0.},offset=-2.1,residual=.01,supportPoints=800,supportCameras=['a','b'])
        south=wall((-2,-2.08),(2,-2.08),'south');south.update(residual=.01,offset=-2.08,supportPoints=700,supportCameras=['a','b'])
        proposals,rejected=core.refine_room_outlines([room],[west,south],scenes())
        self.assertEqual(rejected,[]);self.assertEqual(len(proposals),1);result=proposals[0]
        self.assertEqual(result['id'],room['id']);self.assertEqual(len(result['outline']),4);self.assertEqual(result['openings'],room['openings']);self.assertEqual(result['doorwayCandidates'],room['doorwayCandidates'])
        self.assertLess(result['refinement']['maxCornerShift'],.2);self.assertEqual(result['refinement']['unchangedEdgeLines'],[1,2])
        self.assertEqual(result['outline'][1]['x'],2);self.assertEqual(result['outline'][2],{'x':2.,'z':2.});self.assertEqual(result['outline'][3]['z'],2)

    def test_refinement_never_changes_authored_rooms_or_accepts_large_motion(self):
        room={'id':'room','authored':True,'outline':[{'x':-2,'z':-2},{'x':2,'z':-2},{'x':2,'z':2},{'x':-2,'z':2}]}
        self.assertEqual(core.refine_room_outlines([room],[],scenes())[1][0]['reason'],'authored_state_preserved')
        changed=wall((-2,-2.5),(2,-2.5),'south');changed.update(residual=.01,offset=-2.5,supportPoints=700,supportCameras=['a','b'])
        room['authored']=False;self.assertEqual(core.refine_room_outlines([room],[changed],scenes())[0],[])

if __name__=="__main__":unittest.main()

File: tests/imo3d-depth-architecture.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import childProcess from "node:child_process";
import {EventEmitter} from "node:events";
import {mkdirSync,mkdtempSync,readFileSync,rmSync,writeFileSync} from "node:fs";
import path from "node:path";
import {PassThrough} from "node:stream";
import {applyDepthArchitecture,depthArchitectureSource,runDepthArchitecture} from "../src/lib/imo3d/depth-architecture.ts";
import type {DepthArchitectureResult,DepthArchitectureScene} from "../src/lib/imo3d/depth-architecture.ts";
import type {Plan,PlanRoom} from "../src/lib/imo3d/model.ts";

const rectangle=(x=0,z=0,width=4,height=4)=>[{x,z},{x:x+width,z},{x:x+width,z:z+height},{x,z:z+height}];
const room=():PlanRoom=>({id:"dining",name:"Dining",finish:"wood",outline:rectangle(),openings:[0],doorwayCandidates:[{edge:0,offset:.5,width:.2,confidence:.8,verified:false,pairedRoomId:"hall"}]});
const scenes=():DepthArchitectureScene[]=>[{id:"a",componentId:"main",floor:0,position:{x:1,y:0,z:1},yaw:0},{id:"b",componentId:"main",floor:0,position:{x:2,y:0,z:2},yaw:0}];
const plan=():Plan=>({floor:0,label:"Ground",kind:"estimated",bounds:{minX:0,minZ:0,maxX:4,maxZ:4},walls:[],generatedRooms:[room()],generatedFrom:{method:"image-boundaries",confidence:.6,sceneIds:["a","b"],scale:"camera_height"}});
const area=(outline:{x:number;z:number}[])=>outline.reduce((sum,a,i)=>{const b=outline[(i+1)%outline.length];return sum+a.x*b.z-b.x*a.z;},0)/2;
function proposal(prior=room(),outline=rectangle(0,0,4.1,4.1)){
  return{...structuredClone(prior),outline,refinement:{source:"image_outline_refined_by_multiview_depth",units:"camera_height",verified:false,previousOutline:structuredClone(prior.outline),maxCornerShift:Math.max(...outline.map((p,i)=>Math.hypot(p.x-prior.outline[i].x,p.z-prior.outline[i].z))),areaRatio:area(outline)/area(prior.outline),cameraContainmentPreserved:true,doorEdgeIndicesPreserved:true,unchangedEdgeLines:[0,3],supportedEdgeChanges:[1,2].map(edge=>({edge,wallId:`wall-${edge}`,coverage:.8,beforeResidual:.08,afterResidual:.01,supportPoints:120,supportCameras:["a","b"]}))}};
}
function output(proposed:unknown=proposal()):DepthArchitectureResult{return{version:1,source:depthArchitectureSource,units:"camera_height",groups:[{planIndex:0,floor:0,componentId:"main",scenes:scenes(),status:"estimated",proposedRooms:[proposed],proposalRejections:[]}],warnings:[],outputPaths:[]};}

test("depth architecture admits bounded image-room corrections and preserves exact aperture metadata",()=>{
  const original=plan(),snapshot=structuredClone(original),p=proposal();
  const updated=applyDepthArchitecture([original],scenes(),output({...p,name:"Untrusted rename",finish:"stone"}))[0];
  assert.notEqual(updated,original);assert.deepEqual(original,snapshot);
  assert.deepEqual(updated.generatedRooms?.[0].outline,p.outline);
  assert.equal(updated.generatedRooms?.[0].name,original.generatedRooms?.[0].name);
  assert.deepEqual(updated.generatedRooms?.[0].openings,room().openings);
  assert.deepEqual(updated.generatedRooms?.[0].doorwayCandidates,room().doorwayCandidates);
  assert.ok(updated.walls.length>4);assert.equal(updated.bounds.maxX,4.6);
  assert.equal(updated.generatedFrom?.method,"image-boundaries+multiview-depth-refinement");
  assert.equal(applyDepthArchitecture([updated],scenes(),output())[0],updated,"stale previous outline must not replay");
});

test("depth application rejects malformed evidence, unsupported lines, excessive shifts and altered doors",()=>{
  const invalid=[
    {...proposal(),id:"foreign"},
    {...proposal(),outline:rectangle(0,0,4.3,4.3)},
    {...proposal(),outline:[...proposal().outline,{x:0,z:1}]},
    {...proposal(),openings:[]},
    {...proposal(),doorwayCandidates:[{...room().doorwayCandidates![0],width:.3}]},
    {...proposal(),refinement:{...proposal().refinement,previousOutline:rectangle(0,0,3.9,4)}},
    {...proposal(),refinement:{...proposal().refinement,maxCornerShift:NaN}},
    {...proposal(),refinement:{...proposal().refinement,areaRatio:Infinity}},
    {...proposal(),refinement:{...proposal().refinement,supportedEdgeChanges:[proposal().refinement.supportedEdgeChanges[0]]}},
    {...proposal(),refinement:{...proposal().refinement,supportedEdgeChanges:proposal().refinement.supportedEdgeChanges.map(edge=>({...edge,supportCameras:["a","foreign"]}))}},
    proposal(room(),[{x:.02,z:0},{x:4.1,z:0},{x:4.1,z:4.1},{x:0,z:4.1}]),
    {...proposal(),floor:1}, {...proposal(),componentId:"foreign"},
  ];
  for(const candidate of invalid){const original=plan();assert.equal(applyDepthArchitecture([original],scenes(),output(candidate))[0],original);}
});

test("depth application isolates floors, components, authored rooms and changed camera snapshots",()=>{
  const original=plan();
  for(const mutate of [
    (o:DepthArchitectureResult)=>{o.groups[0].floor=1;},
    (o:DepthArchitectureResult)=>{o.groups[0].componentId="other";},
    (o:DepthArchitectureResult)=>{o.groups[0].scenes[0].position.x+=.01;},
    (o:DepthArchitectureResult)=>{o.groups[0].proposedRooms.push(proposal());},
  ]){const raw=output();mutate(raw);assert.equal(applyDepthArchitecture([original],scenes(),raw)[0],original);}
  for(const authoredRooms of [[],[room()]]){const authored={...plan(),authoredRooms};assert.equal(applyDepthArchitecture([authored],scenes(),output())[0],authored);}
  const upstairs={...plan(),floor:1};const result=applyDepthArchitecture([original,upstairs],scenes(),output());assert.equal(result[1],upstairs);
});

test("depth application recomputes camera containment and added neighbor overlap",()=>{
  const original=plan(),nearEdge={id:"edge",componentId:"main",floor:0,position:{x:3.95,y:0,z:2}};
  assert.equal(applyDepthArchitecture([original],[...scenes(),nearEdge],output(proposal(room(),rectangle(0,0,3.9,3.9))))[0],original);
  const neighbor={...room(),id:"neighbor",outline:rectangle(4,0),openings:[],doorwayCandidates:[]};
  const adjacent={...plan(),generatedRooms:[room(),neighbor]};
  const forged={...proposal(),refinement:{...proposal().refinement,estimatedNewOverlap:0}};
  assert.equal(applyDepthArchitecture([adjacent],scenes(),output(forged))[0],adjacent,"recompute overlap rather than trust the claimed zero");
});

function fixture(){
  const previous=process.env.IMO3D_PYTHON;process.env.IMO3D_PYTHON=process.execPath;
  const root=path.resolve("work");mkdirSync(root,{recursive:true});const directory=mkdtempSync(path.join(root,"depth-architecture-test-"));
  return{directory,options:{points:Array.from({length:100},()=>({x:1,y:1,z:1,r:1,g:2,b:3,confidence:.8,sceneIds:["a","b"]})),scenes:scenes(),plans:[plan()],outputDir:directory},close(){if(previous===undefined)delete process.env.IMO3D_PYTHON;else process.env.IMO3D_PYTHON=previous;assert.ok(path.resolve(directory).startsWith(root+path.sep));rmSync(directory,{recursive:true,force:true});}};
}
function modelProcess(onKill=()=>{}){const child=new EventEmitter() as EventEmitter&{stdout:PassThrough;stderr:PassThrough;kill:()=>boolean};child.stdout=new PassThrough();child.stderr=new PassThrough();child.kill=()=>{onKill();setImmediate(()=>child.emit("close",null));return true;};return child;}

test("depth supervisor isolates floor frames, strips image paths, runs offline and gives retries unique directories",async context=>{
  const f=fixture(),attempts:string[]=[];try{
    context.mock.method(childProcess,"spawn",((_python:string,args:readonly string[],options:childProcess.SpawnOptions)=>{
      assert.equal(options.windowsHide,true);assert.equal(options.env?.HF_HUB_OFFLINE,"1");assert.equal(options.env?.TRANSFORMERS_OFFLINE,"1");
      const input=JSON.parse(readFileSync(args[2],"utf8"));
      assert.equal(new Set(input.scenes.map((s:DepthArchitectureScene)=>s.floor)).size,1);assert.equal(input.pointSamples.length,100);assert.equal(input.scenes.some((s:Record<string,unknown>)=>"path" in s),false);assert.equal(input.priorRooms.length,1);
      attempts.push(path.dirname(args[4]));assert.match(path.basename(attempts.at(-1)!),/^depth-architecture-[\da-f-]+$/);
      const child=modelProcess();setImmediate(()=>{writeFileSync(args[4],JSON.stringify({version:1,units:"camera_height",status:"partial",proposedRooms:[]}));child.emit("close",0);});return child;
    }) as unknown as typeof childProcess.spawn);
    const options={...f.options,scenes:[...scenes().map(s=>({...s,path:"private-do-not-pass.jpg"})),...scenes().map(s=>({...s,id:s.id+"-up",floor:1}))],points:[...f.options.points,...f.options.points.map(p=>({...p,sceneIds:["a-up","b-up"]}))],plans:[plan(),{...plan(),floor:1,generatedFrom:{...plan().generatedFrom!,sceneIds:["a-up","b-up"]}}]};
    const result=await runDepthArchitecture(options);assert.equal(result.groups.length,2);assert.deepEqual(result.groups.map(g=>g.floor),[0,1]);await runDepthArchitecture(options);assert.equal(new Set(attempts).size,4);
  }finally{f.close();}
});

test("depth cancellation kills the child and rejects late partial results and pre-aborted retries",async context=>{
  const f=fixture(),controller=new AbortController();let killed=0;try{
    context.mock.method(childProcess,"spawn",((_python:string,args:readonly string[])=>{const child=modelProcess(()=>killed++);setImmediate(()=>{writeFileSync(args[4],JSON.stringify({version:1,units:"camera_height",status:"estimated",proposedRooms:[proposal()]}));controller.abort();});return child;}) as unknown as typeof childProcess.spawn);
    await assert.rejects(runDepthArchitecture({...f.options,signal:controller.signal}),{name:"AbortError"});assert.equal(killed,1);
    await assert.rejects(runDepthArchitecture({...f.options,signal:controller.signal}),{name:"AbortError"});assert.equal(killed,1);
  }finally{f.close();}
});

test("depth supervisor refuses oversized arrays and oversized or missing output files",async context=>{
  const f=fixture();let calls=0;try{
    context.mock.method(childProcess,"spawn",((_python:string,args:readonly string[])=>{calls++;const child=modelProcess();setImmediate(()=>{if(calls===1)writeFileSync(args[4],Buffer.alloc(8*1024*1024+1,32));child.emit("close",0);});return child;}) as unknown as typeof childProcess.spawn);
    await assert.rejects(runDepthArchitecture({...f.options,points:Array(150001).fill(f.options.points[0])}),/غير صالحة/);assert.equal(calls,0);
    await assert.rejects(runDepthArchitecture(f.options),/output file/);
    await assert.rejects(runDepthArchitecture(f.options),{code:"ENOENT"});assert.equal(calls,2);
  }finally{f.close();}
});
